File: python/21.合并两个有序链表.py
#
# @lc app=leetcode.cn id=21 lang=python3
#
# [21] 合并两个有序链表
#


# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # 递归
    def mergeTwoLists(self, list1: Optional[ListNode],
                      list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        elif not list2:
            return list1
        elif list1.val > list2.val:
            list2.next = self.mergeTwoLists(list2.next, list1)
            return list2
        else:
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1

    # 迭代
    def mergeTwoLists1(self, list1: Optional[ListNode],
                       list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, None)
        cur = dummy
        while list1 and list2:
            if list1.val > list2.val:
                cur.next = list2
                list2 = list2.next
            else:
                cur.next = list1
                list1 = list1.next
            cur = cur.next

        cur.next = list1 if not list2 else list2

        return dummy.next


# @lc code=end

File: cpp/21.合并两个有序链表.cpp
/*
 * @lc app=leetcode.cn id=21 lang=cpp
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2)
    {
        return dfs(list1, list2);
    }

    ListNode* dfs(ListNode* list1, ListNode* list2) {
        if (!list1 && !list2) {
            return nullptr;
        }

        ListNode* ret = nullptr;
        if (!list1 || !list2) {
            ret = !list1 ? list2 : list1;
            return ret;
        }

        if (list1->val > list2->val) {
            ret = list2;
            ret->next = dfs(list1, list2->next);
        } else {
            ret = list1;
            ret->next = dfs(list1->next, list2);
        }

        return ret;
    }

    ListNode* bfs(ListNode* list1, ListNode* list2)
    {
        ListNode* dummy = new ListNode(0);
        ListNode* cur   = dummy;

        ListNode* ptr1 = list1;
        ListNode* ptr2 = list2;

        while (ptr1 && ptr2) {
            if (ptr1->val > ptr2->val) {
                cur->next = ptr2;
                ptr2      = ptr2->next;
            }
            else {
                cur->next = ptr1;
                ptr1      = ptr1->next;
            }
            cur = cur->next;
        }

        cur->next = ptr1 ? ptr1 : ptr2;
        ListNode* ans = dummy->next;
        delete dummy;
        return ans;
    }
};
// @lc code=end
